runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java

```diff
@@ -23,43 +26,136 @@
         Message>> {
 
     private final HashMap<String,
-            Index.Spec<Message, CorfuRecord<Message, Message>, ? extends Comparable<?>>>
+            Index.Spec<Message, CorfuRecord<Message, Message>, ?>>
             indices = new HashMap<>();
 
     ProtobufIndexer(Message payloadSchema) {
-        payloadSchema.getDescriptorForType().getFields().forEach(this::registerIndices);
+        payloadSchema.getDescriptorForType().getFields().forEach(field -> registerIndices(field, payloadSchema));
     }
 
-    private <T extends Comparable<T>> Index.Spec<Message, CorfuRecord<Message, Message>, ? extends Comparable<?>>
+    private <T> Index.Spec<Message, CorfuRecord<Message, Message>, ?>
     getIndex(String indexName, FieldDescriptor fieldDescriptor) {
-
         return new Index.Spec<>(
                 () -> indexName,
                 (Index.Function<Message, CorfuRecord<Message, Message>, T>)
                         (key, val) -> ClassUtils.cast(val.getPayload().getField(fieldDescriptor)));
     }
 
-    void registerIndices(final Descriptors.FieldDescriptor fieldDescriptor) {
+    private <T> Index.Spec<Message, CorfuRecord<Message, Message>, ?>
+    getComplexIndex(String indexName) {
+        return new Index.Spec<>(
+                () -> indexName,
+                (Index.Function<Message, CorfuRecord<Message, Message>, T>)
+                        (key, val) -> {
+                            // Get all nested fields, '.' separated
+                            String[] nestedFields = indexName.split("\\.");
+
+                            if (nestedFields.length > 0) {
+                                Message subMessage = val.getPayload();
+
+                                for (int i =0; i < nestedFields.length; i++) {
+                                    try {
+                                        subMessage = (Message) subMessage.getField(
```

Comment:
So this is meant to go 2 levels right, not N levels? Or am I missing something?
2 levels is actually fine too!

Reply:
It is bounded to nestedFields.length, so it can actually go to N levels. It will return outside of the for loop, unless a cast exception is encountered and it's in the last level.